File: tansuo-bytedance/34LRU缓存机制.java
class Node{
    public int key,val;
    public Node next,pre;
    public Node(int x,int y){
        this.key = x;
        this.val = y;
    }
}

class DoulbeList{
    private Node head,tail;//头尾结点
    private int size;//链表元素数
    
    public DoubleList(){
        head = new Node(0,0);
        tail = new Node(0, 0);
        head.next = tail;
        tail.prev = head;
        size = 0;
    }
    
    //在链表头部添加节点
    public void addFirst(Node x){
        x.next = head.next;
        x.pre = head;
        head.next.pre = x;
        head.next = x;
        size++;
    }
    
    //删除链表中的x结点（x一定存在）
    public void remove(Node x){
        x.pre.next = x.next;
        x.next.pre = x.pre;
        size--;
    }
    
    // 删除链表中最后一个节点，并返回该节点
    public Node removeLast() {
        if (tail.prev == head)//tail是虚节点
            return null;
        Node last = tail.prev;
        remove(last);
        return last;
    }
    
    // 返回链表长度
    public int size() { return size; }
}

class LRUCache {
    private HashMap<Integer,Node> map; //key->Node(key,val)
    private DoubleList cache;// Node(k1,v1)<->Node(k2,v2)
    private int cap;//最大容量
    
    public LRUCache(int capacity) {
        this.cap = capacity;
        map = new HashMap<>();
        cache = new DoubleList<>();
    }
    
    
    public int get(int key) {
        if (!map.containsKey(key))
            return -1;
        int val = map.get(key).val;
        // 利用 put 方法把该数据提前
        put(key, val);
        return val;
    }
    
    public void put(int key, int val) {
        // 先把新节点 x 做出来
        Node x = new Node(key, val);
        
        if (map.containsKey(key)) {
            // 删除旧的节点，新的插到头部
            cache.remove(map.get(key));
            cache.addFirst(x);
            // 更新 map 中对应的数据
            map.put(key, x);
        } else {
            if (cap == cache.size()) {
                // 删除链表最后一个数据
                Node last = cache.removeLast();
                map.remove(last.key);
            }
            // 直接添加到头部
            cache.addFirst(x);
            map.put(key, x);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: tansuo-bytedance/34LRU缓存机制.py
# 运用你所掌握的数据结构，设计和实现一个 LRU (最近最少使用) 缓存机制。它应该支持以下操作： 获取数据 get 和 写入数据 put 。 
# 
#  获取数据 get(key) - 如果关键字 (key) 存在于缓存中，则获取关键字的值（总是正数），否则返回 -1。 
# 写入数据 put(key, value) - 如果关键字已经存在，则变更其数据值；如果关键字不存在，则插入该组「关键字/值」。当缓存容量达到上限时，它应该在
# 写入新数据之前删除最久未使用的数据值，从而为新的数据值留出空间。 
# 
#  
# 
#  进阶: 
# 
#  你是否可以在 O(1) 时间复杂度内完成这两种操作？ 
# 
#  
# 
#  示例: 
# 
#  LRUCache cache = new LRUCache( 2 /* 缓存容量 */ );
# 
# cache.put(1, 1);
# cache.put(2, 2);
# cache.get(1);       // 返回  1
# cache.put(3, 3);    // 该操作会使得关键字 2 作废
# cache.get(2);       // 返回 -1 (未找到)
# cache.put(4, 4);    // 该操作会使得关键字 1 作废
# cache.get(1);       // 返回 -1 (未找到)
# cache.get(3);       // 返回  3
# cache.get(4);       // 返回  4
#  
#  Related Topics 设计 
#  👍 759 👎 0


# leetcode submit region begin(Prohibit modification and deletion)

class DLinkedNode:
    def __init__(self,key=0,value=0):
        self.key=key
        self.value=value
        self.prev=None
        self.next=None

class LRUCache(object):

    def __init__(self, capacity:int):
        """
        :type capacity: int
        """
        self.cache=dict()
        # 使用伪头部和伪尾部结点
        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        self.head.next=self.tail
        self.tail.prev=self.head
        self.capacity=capacity
        self.size=0


    def get(self, key:int)->int:
        """
        :type key: int
        :rtype: int
        """
        if key not in self.cache:
            return -1
        # 如果key存在，先通过哈希表定位，再移动到头部
        node = self.cache[key]
        self.moveToHead(node)
        return node.value

    def put(self, key:int , value:int ):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key not in self.cache:
            # 如果key不存在，创建一个新节点
            node = DLinkedNode(key,value)
            # 添加到hash表
            self.cache[key] = node
            #添加到双向链表的头部
            self.addToHead(node)
            self.size+=1
            if self.size >self.capacity:
                #如果超出容量，删除双向链表的尾结点
                removed = self.removeTail()
                #删除哈希表中对应的项
                self.cache.pop(removed.key)
                self.size-=1
        else:
            # 如果 key 存在，先通过哈希表定位，再修改 value，并移到头部
            node = self.cache[key]
            node.value = value
            self.moveToHead(node)

    def addToHead(self,node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node

    def removeNode(self,node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def moveToHead(self,node):
        self.removeNode(node)
        self.addToHead(node)

    def removeTail(self):
        node = self.tail.prev
        self.removeNode(node)
        return node



# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# leetcode submit region end(Prohibit modification and deletion)
